src/utils/collections.zig

```diff
@@ -486,6 +486,121 @@ pub fn orderSlices(
     return if (a.len == b.len) .eq else if (a.len > b.len) .gt else .lt;
 }
 
+/// Stores a range of values centered at a particular index that may
+/// change with respect to time.
+///
+/// This is useful when you are tracking multiple different states
+/// that occur in a sequence, and they each have a predefined lifetime.
+///
+/// For example, let's say you need to create a new data structure to
+/// represent every Epoch. At any point in time, you'd like store to
+/// the struct for the current Epoch as well as the next Epoch and
+/// the prior Epoch. You can create a Window with a size of 3 centered
+/// at the current Epoch. It will allow you to store all three of the
+/// Epochs you care about. When transitioning Epochs, call `realign`
+/// to adjust the center to the new Epoch. This will delete the state
+/// from the old Epoch out of the Window, and open up a slot for you
+/// to insert the next Epoch.
+pub fn Window(T: type) type {
+    return struct {
+        state: []?T,
+        center: usize,
+        offset: usize,
+
+        const Self = @This();
+
+        pub fn init(allocator: Allocator, len: usize, start: usize) !Self {
+            const state = try allocator.alloc(?T, len);
+            @memset(state, null);
+            return .{
+                .state = state,
+                .center = start,
+                .offset = len - (start % len),
+            };
+        }
+
+        pub fn deinit(self: Self, allocator: Allocator) void {
+            allocator.free(self.state);
+        }
+
+        /// Inserts the item into the Window, as long as its index
+        /// is within the current allowed bounds of the Window.
+        pub fn put(self: *Self, index: usize, item: T) error{OutOfBounds}!?T {
+            if (!self.isInRange(index)) {
```

Comment:
NIT: tend to prefer single line if it fits within out col limits but not that fussed

Reply:
Typically I'll agree with collapsing if-statements when they can fit on a single line, but if it's a return or a break, I give it a second thought. This was a conscious decision to break it into multiple lines.

I feel that the exit points of a function are critical to understand how the function works, and they should be given extra emphasis compared to other code. Putting it at the end of the line makes it less obvious. I find it easier to comprehend a function's overall flow when all the return statements are isolated on their own lines.

I think the benefit of reducing lines of code is not worth the clarity tradeoff, unless there are many instances of the same pattern, and the repetition makes the pattern obvious. For example, if the function starts with five lines of code that all end in `orelse return null`, then I think it's overkill to allocate extra lines of code for that. But if it's just a single return, I try to highlight it.

Reply:
I tend to find syntax highlighting sufficient for this purpose but also don't feel that strongly to happy to leave it as is